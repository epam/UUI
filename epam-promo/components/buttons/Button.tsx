import * as uui from '@epam/uui';
import { createSkinComponent, devLogger } from '@epam/uui-core';
import { FillStyle } from '../types';
import css from './Button.module.scss';

/** Defines component color. */
type ButtonColor = 'blue' | 'green' | 'red' | 'gray50' | 'gray' | uui.ButtonProps['color'];

type ButtonMods = {
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: FillStyle;
    /**
     * Defines component color.
     * @default 'blue'
     */
    color?: ButtonColor;
    /**
     * Defines component size.
     * @default '36'
     */
    size?: uui.ButtonProps['size'];
};

const mapFill: Record<FillStyle, uui.ButtonProps['fill']> = {
    solid: 'solid',
    white: 'outline',
    light: 'ghost',
    none: 'none',
};

/** Represents the properties of a Button component. */
export type ButtonProps = uui.ButtonCoreProps & ButtonMods;

export const Button = /* @__PURE__ */createSkinComponent<uui.ButtonProps, ButtonProps>(
    uui.Button as any, // TODO: remove it when BaseButton inheritance will be reworked
    (props) => {
        if (__DEV__) {
            devLogger.warnAboutDeprecatedPropValue<ButtonProps, 'color'>({
                component: 'Button',
                propName: 'color',
                propValue: props.color,
                propValueUseInstead: 'gray',
                condition: () => ['gray50'].indexOf(props.color) !== -1,
            });
        }
        return {
            fill: mapFill[props.fill] || mapFill.solid,
        } as any; // TODO: need new helper to rewrite types
    },
    (props) => [
        ['42', '48'].includes(props.size) && css.uppercase,
    ],
);
