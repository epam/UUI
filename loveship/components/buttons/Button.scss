// Border style
.style-round {
    &.size-18 { border-radius: 10px; }
    &.size-24 { border-radius: 12px; }
    &.size-30 { border-radius: 15px; }
    &.size-36 { border-radius: 18px; }
    &.size-42 { border-radius: 21px; }
    &.size-48 { border-radius: 24px; }
}