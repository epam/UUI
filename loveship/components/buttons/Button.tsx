import { FillStyle, ControlShape, EpamPrimaryColor } from '../types';
import * as uui from '@epam/uui';
import { createSkinComponent, devLogger } from '@epam/uui-core';
import css from './Button.module.scss';

const defaultSize = '36';

/** Defines component color. */
type ButtonColor = EpamPrimaryColor | 'white' | 'night500' | 'night600' | 'gray' | uui.ButtonProps['color'];

type ButtonMods = {
    /**
     * Defines component color.
     * @default "sky"
     */
    color?: ButtonColor;
    /**
     * Defines component size.
     * @default '36'
     */
    size?: uui.ButtonProps['size'] | '18';
    /**
     * Defines component shape.
     * @default 'square'
     */
    shape?: ControlShape;
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: FillStyle;
};

const mapFill: Record<FillStyle, uui.ButtonProps['fill']> = {
    solid: 'solid',
    white: 'outline',
    light: 'ghost',
    none: 'none',
};

/** Represents the properties of a Button component. */
export type ButtonProps = uui.ButtonCoreProps & ButtonMods;

function applyButtonMods(mods: ButtonProps) {
    return [
        `uui-size-${mods.size || defaultSize}`,
        css['style-' + (mods.shape || 'square')],
    ];
}

export const Button = /* @__PURE__ */createSkinComponent<uui.ButtonProps, ButtonProps>(
    uui.Button as any, // TODO: remove it when BaseButton inheritance will be reworked
    (props) => {
        if (__DEV__) {
            devLogger.warnAboutDeprecatedPropValue<ButtonProps, 'color'>({
                component: 'Button',
                propName: 'color',
                propValue: props.color,
                propValueUseInstead: 'gray',
                condition: () => ['night500', 'night600'].indexOf(props.color) !== -1,
            });
        }
        return {
            fill: mapFill[props.fill] || mapFill.solid as any,
        };
    },
    applyButtonMods,
);
