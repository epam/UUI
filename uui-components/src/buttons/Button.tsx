import * as React from 'react';
import { ButtonCoreProps, Icon, uuiElement, uuiMarkers, CX, IHasRawProps, cx, IHasForwardedRef } from '@epam/uui-core';
import { IconContainer } from '../layout';
import { ButtonBase } from './ButtonBase';
import css from './Button.scss';

export interface ButtonProps
    extends ButtonCoreProps,
        IHasRawProps<React.ButtonHTMLAttributes<HTMLButtonElement>>,
        IHasForwardedRef<HTMLButtonElement | HTMLAnchorElement> {
    /** Icon for clear value button (usually cross) */
    clearIcon?: Icon;

    /** CSS classes to put on the caption */
    captionCX?: CX;
}

export class Button extends ButtonBase<ButtonProps> {
    constructor(props: ButtonProps) {
        super(props);
    }

    getClassName() {
        return [css.container];
    }

    getChildren() {
        return [
            this.props.isDropdown && this.props.dropdownIconPosition === 'left' && (
                <IconContainer key="dropdown-icon-left" icon={this.props.dropdownIcon} flipY={this.props.isOpen} />
            ),
            this.props.icon && this.props.iconPosition !== 'right' && (
                <IconContainer key="icon-left" icon={this.props.icon} onClick={!this.props.isDisabled ? this.props.onIconClick : undefined} />
            ),
            this.props.count !== undefined && this.props.count !== null && this.props.countPosition !== 'right' && (
                <div key="count" className={cx(uuiElement.count)}>
                    {this.props.count}
                </div>
            ),
            this.props.caption && (
                <div key="caption" className={cx(uuiElement.caption, this.props.captionCX)}>
                    {this.props.caption}
                </div>
            ),
            this.props.count !== undefined && this.props.count !== null && this.props.countPosition === 'right' && (
                <div key="count" className={cx(uuiElement.count)}>
                    {this.props.count}
                </div>
            ),
            !this.props.caption && this.props.placeholder && (
                <div key="placeholder" className={cx(uuiElement.caption, uuiElement.placeholder, this.props.captionCX)}>
                    {this.props.placeholder}
                </div>
            ),
            this.props.icon && this.props.iconPosition === 'right' && (
                <IconContainer key="icon-right" icon={this.props.icon} onClick={!this.props.isDisabled ? this.props.onIconClick : undefined} />
            ),
            this.props.isDropdown && this.props.dropdownIconPosition !== 'left' && (
                <IconContainer key="dropdown-icon-right" icon={this.props.dropdownIcon} flipY={this.props.isOpen} />
            ),
            this.props.onClear && !this.props.isDisabled && (
                <IconContainer key="clear-icon" cx={uuiMarkers.clickable} icon={this.props.clearIcon} onClick={this.props.onClear} />
            ),
        ];
    }
}
