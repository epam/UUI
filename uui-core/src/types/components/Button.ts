import * as React from 'react';
import {
    IBasicPickerToggler,
    ICanRedirect,
    IClickable,
    IDisableable,
    IDropdownToggler,
    IHasCaption,
    IHasCX,
    IHasIcon,
    IHasPlaceholder,
    IAnalyticableClick,
    IHasTabIndex,
    IHasRawProps,
} from '../props';
import { Icon } from '../objects';

export interface ButtonBaseCoreProps
    extends IHasCX,
        IClickable,
        ICanRedirect,
        IDisableable,
        IHasIcon,
        IAnalyticableClick,
        IHasTabIndex,
        IHasRawProps<React.ButtonHTMLAttributes<HTMLButtonElement>> {}

export interface ButtonCoreProps extends ButtonBaseCoreProps, IHasCaption, IBasicPickerToggler, IDropdownToggler, IHasPlaceholder {
    /** Icon for drop-down toggler */
    dropdownIcon?: Icon;

    /** Position of the dropdown icon ('left' of 'right') */
    dropdownIconPosition?: 'left' | 'right';

    countPosition?: 'left' | 'right';
    count?: number | null;
}

export interface ButtonSemanticProps {
    type?: 'success' | 'cancel' | 'action';
}
