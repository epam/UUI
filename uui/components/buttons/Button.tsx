import * as uuiComponents from '@epam/uui-components';
import { devLogger, withMods } from '@epam/uui-core';
import { ControlSize } from '../types';
import { systemIcons } from '../../icons/icons';
import css from './Button.module.scss';

const DEFAULT_SIZE = '36';

type ButtonMods = {
    /**
     * Defines component size.
     * @default '36'
     */
    size?: ControlSize | '18';
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: 'solid' | 'outline' | 'ghost' | 'none';
    /**
     * Defines component color.
     * @default 'primary'
     */
    color?: 'accent' | 'primary' | 'critical' | 'secondary' | 'neutral';
};

/** Represents the 'Core properties' for the Button component. */
export type ButtonCoreProps = uuiComponents.ButtonProps;

/** Represents the props for a Button component. */
export type ButtonProps = ButtonMods & ButtonCoreProps;

function applyButtonMods(mods: ButtonProps) {
    return [
        css.root,
        'uui-button',
        `uui-fill-${mods.fill || 'solid'}`,
        `uui-color-${mods.color || 'primary'}`,
        `uui-size-${mods.size || DEFAULT_SIZE}`,
    ];
}

export const Button = /* @__PURE__ */withMods<uuiComponents.ButtonProps, ButtonMods>(
    uuiComponents.Button,
    applyButtonMods,
    (props) => {
        if (__DEV__ && props.captionCX) {
            devLogger.warn('Button: Property \'captionCX\' is deprecated and will be removed in the future release. Please use \'cx\' prop to access caption styles and use cascading to change the styles for the \'uui-caption\' global class');
        }

        return {
            dropdownIcon: systemIcons.foldingArrow,
            clearIcon: systemIcons.clear,
        };
    },
);
