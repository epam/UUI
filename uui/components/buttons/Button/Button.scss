@use '../../../assets/styles/index' as *;

.root {
    &:global(.uui-button-box) {
        border-style: solid;
        justify-content: center;
        min-width: inherit;
        border-radius: var(--uui-btn-border-radius);
    }

    :global(.uui-caption) {
        @include overflow-ellipsis();
        min-width: 0;
        text-align: center;
        flex: 1 1 auto;
        font-family: var(--font-semibold);
        font-weight: normal;
        user-select: none;
    }
}

@mixin button-disables-styles($bg-color, $border-color) {
    background-color: $bg-color;
    color: var(--uui-button-caption-disabled);
    fill: var(--uui-button-caption-disabled);
    border-color: $border-color;
}

.mode-solid {
    background-color: var(--uui-btn-bg);
    color: var(--uui-btn-caption);
    fill: var(--uui-btn-caption);
    border-color: var(--uui-btn-bg);

    &:global(.-clickable) {
        &:hover {
            background-color: var(--uui-btn-bg-hover);
            border-color: var(--uui-btn-bg-hover);
        }

        &:focus {
            background-color: var(--uui-btn-bg-hover);
            border-color: var(--uui-btn-bg-hover);
            box-shadow: var(--uui-btn-focus-shadow);
            outline: var(--uui-btn-focus-outline);
        }

        &:active {
            background-color: var(--uui-btn-bg-active);
            border-color: var(--uui-btn-bg-active);
        }
    }

    &:global(.uui-disabled) {
        @include button-disables-styles(var(--uui-btn-bg-disabled), var(--uui-btn-border-disabled));
    }

    &:global(.-clickable), &:hover, &:active, &:focus {
        &:global(.uui-disabled) {
            @include button-disables-styles(var(--uui-btn-bg-disabled), var(--uui-btn-border-disabled));
        }
    }
}

.mode-outline {
    background-color: var(--uui-button-bg-halftone);
    color: var(--uui-btn-caption-halftone);
    fill: var(--uui-btn-caption-halftone);
    border-color: var(--uui-btn-border);

    &:global(.-clickable) {
        &:hover, &:focus {
            background-color: var(--uui-btn-bg-halftone-hover);
            color: var(--uui-btn-caption-halftone);
            fill: var(--uui-btn-caption-halftone);
            border-color: var(--uui-btn-border);
        }

        &:active {
            background-color: var(--uui-btn-bg-halftone-active);
            color: var(--uui-btn-caption-halftone);
            fill: var(--uui-btn-caption-halftone);
            border-color: var(--uui-btn-border);
        }
    }

    &:global(.uui-disabled) {
        @include button-disables-styles(var(--uui-button-bg-halftone), var(--uui-btn-border-disabled));
    }

    &:global(.-clickable), &:hover, &:active, &:focus {
        &:global(.uui-disabled) {
            @include button-disables-styles(var(--uui-button-bg-halftone), var(--uui-btn-border-disabled));
        }
    }
}

.mode-none {
    background-color: transparent;
    color: var(--uui-btn-caption-halftone);
    fill: var(--uui-btn-caption-halftone);
    border-color: var(--uui-btn-border);

    &:global(.-clickable) {
        &:hover, &:focus {
            background-color: var(--uui-btn-bg-halftone-hover);
            color: var(--uui-btn-caption-halftone);
            fill: var(--uui-btn-caption-halftone);
            border-color: var(--uui-btn-border);
        }

        &:active {
            background-color: var(--uui-btn-bg-halftone-active);
            color: var(--uui-btn-caption-halftone);
            fill: var(--uui-btn-caption-halftone);
            border-color: var(--uui-btn-border);
        }
    }

    &:global(.uui-disabled) {
        @include button-disables-styles(transparent, var(--uui-btn-border-disabled));
    }

    &:global(.-clickable), &:hover, &:active, &:focus {
        &:global(.uui-disabled) {
            @include button-disables-styles(transparent, var(--uui-btn-border-disabled));
        }
    }
}

.mode-ghost {
    background-color: transparent;
    color: var(--uui-btn-caption-halftone);
    fill: var(--uui-btn-caption-halftone);
    border-color: transparent;

    &:global(.-clickable) {
        &:hover, &:focus {
            background-color: var(--uui-btn-bg-halftone-hover);
            border-color: var(--uui-btn-bg-halftone-hover);
        }

        &:active {
            background-color: var(--uui-btn-bg-halftone-active);
            border-color: var(--uui-btn-bg-halftone-active);
        }
    }


    &:global(.uui-disabled) {
        @include button-disables-styles(transparent, transparent);
    }


    &:global(.-clickable), &:hover, &:active, &:focus {
        &:global(.uui-disabled) {
            @include button-disables-styles(transparent, transparent);
        }
    }
}

//Sizes
.size-24 {
    @include button-layout(24px, 1px);
}

.size-30 {
    @include button-layout(30px, 1px);
}

.size-36 {
    @include button-layout(36px, 1px);
}

.size-42 {
    @include button-layout(42px, 1px);

    :global(.uui-caption) {
        font-family: var(--font-primary);
        text-transform: uppercase;
    }
}

.size-48 {
    @include button-layout(48px, 1px);

    :global(.uui-caption) {
        font-family: var(--font-primary);
        text-transform: uppercase;
    }
}