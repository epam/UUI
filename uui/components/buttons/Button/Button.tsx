import { Button as uuiButton, ButtonProps } from '@epam/uui-components';
import { withMods } from '@epam/uui-core';
import { ControlSize, ButtonMode } from '../../types';
import { systemIcons } from '../../../icons/icons';
import css from './Button.scss';

export type ButtonColor = 'accent' | 'primary' | 'secondary' | 'negative';
export const allButtonColors: ButtonColor[] = ['accent', 'primary', 'secondary', 'negative'];

const defaultSize = '36';

export interface ButtonMods {
    size?: ControlSize | '18';
    mode?: ButtonMode;
    color?: ButtonColor;
}

export type UUIButtonProps = ButtonMods & ButtonProps;

export function applyButtonMods(mods: UUIButtonProps) {
    return [
        'button-vars',
        `button-${mods.color || 'primary'}`,
        css.root,
        css[`size-${mods.size || defaultSize}`],
        css[`mode-${mods.mode || 'solid'}`],
    ];
}

export const Button = withMods<ButtonProps, ButtonMods>(
    uuiButton,
    applyButtonMods,
    (props) => ({
        dropdownIcon: systemIcons[props.size || defaultSize].foldingArrow,
        clearIcon: systemIcons[props.size || defaultSize].clear,
    }),
);
