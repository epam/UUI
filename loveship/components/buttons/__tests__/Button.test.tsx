import React from 'react';
import renderer from 'react-test-renderer';
import { Button } from '../Button';
import { ReactComponent as AcceptIcon } from '../../icons/accept-12.svg';

describe('Button', () => {
    it('should be rendered correctly', () => {
        const tree = renderer.create(<Button />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with props', () => {
        const tree = renderer
            .create(
                <Button
                    color="sun"
                    fill="white"
                    caption="Click me"
                    onClick={jest.fn()}
                    onClear={jest.fn()}
                    icon={AcceptIcon}
                    isDisabled={false}
                    isDropdown={true}
                    count={10}
                    iconPosition="right"
                    isOpen
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
