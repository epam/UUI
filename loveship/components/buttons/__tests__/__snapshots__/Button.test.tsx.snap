// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be rendered correctly 1`] = `
<button
  className="container uui-button-box uui-enabled root size-36 color-sky fill-solid style-square"
  href={null}
  onClick={[Function]}
  role="button"
  tabIndex={-1}
  type="button"
/>
`;

exports[`Button should be rendered correctly with props 1`] = `
<button
  aria-disabled={false}
  className="container uui-button-box uui-enabled -clickable root size-36 color-sun fill-white style-square"
  href={null}
  onClick={[Function]}
  role="button"
  tabIndex={0}
  type="button"
>
  <div
    className="uui-count"
  >
    10
  </div>
  <div
    className="uui-caption line-height-18 font-size-14 v-padding-8"
  >
    Click me
  </div>
  <div
    className="container uui-icon uui-enabled"
    style={Object {}}
  >
    <svg
      className=""
    />
  </div>
  <div
    className="container uui-icon uui-enabled"
    style={Object {}}
  >
    <svg
      className="flipY"
    />
  </div>
  <div
    className="container uui-icon uui-enabled -clickable -clickable"
    onClick={[MockFunction]}
    style={Object {}}
  >
    <svg
      className=""
    />
  </div>
</button>
`;
